packages/accounts-controller/CHANGELOG.md

```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [20.0.0]
+
+### Changed
+
+- **BREAKING:** Bump peer dependency `@metamask/keyring-controller` from `^18.0.0` to `^19.0.0` ([#4915](https://github.com/MetaMask/core/pull/4956))
+- **BREAKING:** Bump `@metamask/keyring-api` from `^8.1.3` to `^10.1.0` ([#4948](https://github.com/MetaMask/core/pull/4948))
+  - Bump the peer dependency `@metamask/providers` from `^17.2.0` to `^18.1.0`.
```

Comment:
Hmm, if we are asking the consumer of `@metamask/accounts-controller` to upgrade their dependency on `@metamask/providers`, I'm not sure that's correct. In this case `@metamask/accounts-controller` is the consumer of `@metamask/keyring-api`, so _it_ should depend on `@metamask/providers`, not the client. Yarn even says so when you run `yarn install`:

```
➤ YN0002: │ @metamask/accounts-controller@workspace:packages/accounts-controller [cbb75] doesn't provide @metamask/providers (pa99fc), requested by @metamask/keyring-api.
➤ YN0002: │ @metamask/accounts-controller@workspace:packages/accounts-controller [d7116] doesn't provide @metamask/providers (pcc967), requested by @metamask/keyring-api.
```

Since these peer dependency warnings are also a problem with other controllers that depend on `keyring-api`, should we quickly fix those controllers before this release goes out? I can submit a PR if need be.

In the meantime, maybe we don't need the `BREAKING:` here? What do you think @ccharly and @mikesposito?

``` suggestion
```

Reply:
So there's 3 ways to fix this I believe?
1. Either we "forward" the peer dependency (which will be resolved by the client in this case)
2. Or we have a "direct" dependency in every controllers that depends on the `keyring-controller`
3. Lastly, I think we could have that "direct" depedency on the `keyring-controller` itself

I'm not sure which one is better... We used to have the `@metamask/providers` as a dependency on the `keyring-api`: https://github.com/MetaMask/keyring-api/pull/282/files

Having it as a peer dependency makes sense but definitely adds a bit of complexity IMO 😅

Reply:
This would also mean that `keyring-controller` needs to have `@metamask/providers` as a direct dependency, right? 

Though, I see that `keyring-api` [only pulls types](https://github.com/MetaMask/accounts/blob/8103bff727d3afce89bff0ecb2ffb6f1aa2efabc/packages/keyring-api/src/KeyringSnapRpcClient.ts#L1-L1) from `@metamask/providers`, while from what I understand it is the client that [constructs the actual provider instance](https://github.com/MetaMask/metamask-extension/blob/02d52b49c06a6b1d8b65e4ee4450aa7db2357220/app/scripts/inpage.js#L37-L37) - so we should apply the same criteria that we use for controllers that depends on other controllers: they all rely on the fact that the it will be the client to install the dependency and use that specific version with the packages that rely on it, rather than letting each single package install a different version of the peer dep

does this make sense?

Reply:
> so we should apply the same criteria that we use for controllers that depends on other controllers

If this is the case then maybe `@metamask/providers` should just be a `devDependency` for packages like `keyring-controller` and `accounts-controller`?

Reply:
Actually, since also `@metamask/keyring-controller` also uses only types from `@metamask/keyring-api`, should `@metamask/keyring-api` be a `devDependency` + `peerDependency` of `@metamask/keyring-controller` rather than a production one?

Reply:
Strictly speaking, if we are using a dependency only for types, then it should be in `dependencies` and there is no need for it to be in `peerDependencies`. (In the past I was confused about this but I believe we've settled on this now.) Unless there is something special about `@metamask/providers` we should be considering, just using `dependencies` "should" be okay because if different versions of a package that ships with types are present in the dependency tree — say, a client uses one version of `@metamask/providers` and upgrades a package in the dependency tree that also uses `@metamask/providers` and those versions are different — then the incompatibility will show up as a build error, forcing the client to upgrade.

That said, the peer dependency warnings we're getting in `core` about `@metamask/providers` have been there since `@metamask/keyring-api` v8.0.0, and I don't think upgrading to v10.1.0 adds more. So, I guess this PR is okay for now. At the same time, I think it's okay to mention that the `@metamask/keyring-api` upgrade is breaking. A consumer that imports `@metamask/providers` and also uses `@metamask/accounts-controllers` _will_ have to upgrade, so that's a good callout. Maybe we can just clarify the language a bit as it's a bit unclear right now. I will make a suggestion on that:

``` suggestion
- **BREAKING:** Bump `@metamask/keyring-api` from `^8.1.3` to `^10.1.0` ([#4948](https://github.com/MetaMask/core/pull/4948))
  - If you are depending on `@metamask/providers` directly, you will need to upgrade to 18.1.0.
```

Past this, we can correct the `@metamask/providers` dependency and at the same time address the peer dependency warnings in `core` a future release.

Does that sound like a good plan?

Reply:
Sounds good to me yes! And I'm also ok with the suggestions.

Also, FYI we might split the `keyring-api` "somedays" 😅 and splitting the "clients" part of the `keyring-api` (which is the one that requires the `@metamask/providers`) into a `keyring-api-client` might solves a lot of issue with those peer dependencies.